### PR TITLE
Feature complex columns

### PR DESCRIPTION
**Pull request by ronanstokes-db**
## Proposed changes

Add support for use of complex columns such as array, map and struct type columns.

Changes include:
- allowing use of constructs such as "map<string, int>" in column type definition
- defaulting columns of type array, map and struct to null unless `expr` attribute is populated
- adding test cases for use of map, struct and array columns
- validating reported issues around use of array and struct valued columns

## Types of changes

What types of changes does your code introduce to dbldatagen?
_Put an `x` in the boxes that apply_

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Change to tutorials, tests or examples
- [ ] Non code change (readme, images or other non-code assets)
- [x] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. 
If you're unsure about any of them, don't hesitate to ask. We're here to help! 
This is simply a reminder of what we are going to look for before merging your code._

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] Submission does not reduce code coverage numbers
- [ ] Submission does not increase alerts or messages from LGTM

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you 
did and what alternatives you considered, etc...

### CODE REVIEW

File: dbldatagen/schema_parser.py [hunk 1]
```diff
@@ -6,83 +6,259 @@
...
+        elif first_token == "double":
             retval = DoubleType()
-        elif s == "float":
+        elif first_token == "float":
             retval = FloatType()
-        elif s == "date":
+        elif first_token == "date":
             retval = DateType()
-        elif s == "short":
+        elif first_token in ["short", "smallint"]:
```
Comment (howardwu-db) [inline dbldatagen/schema_parser.py]: should be first_token == "smallint"

File: dbldatagen/schema_parser.py [hunk 2]
```diff
@@ -6,83 +6,259 @@
...
             retval = ShortType()
-        elif s == "byte":
+        elif first_token in ["byte", "tinyint"]:
```
Comment (howardwu-db) [inline dbldatagen/schema_parser.py]: should be first_token == "tinyint"

**Review by howardwu-db — CHANGES_REQUESTED** [review]

File: dbldatagen/schema_parser.py [hunk 3]
```diff
@@ -6,83 +6,259 @@
...
+
+        See the package pyparsing for details of how the parser mechanism works
+        `https://pypi.org/project/pyparsing/`
         """
-        assert type_string is not None, "`type_string` must be specified"
+        # define grammar for parsing type definitions of the form "INT", "decimal(10,4)"
+        # and "map<string, string>" etc
+        if cls._type_parser is None:
+            int_keyword = pp.CaselessKeyword("int") ^ pp.CaselessKeyword("integer")
+            int_keyword.setParseAction(lambda s: "int")
```
[the remark this answers is not in the document]
Comment (ronanstokes-db) [reply dbldatagen/schema_parser.py:57]: thats in a much later version

version 2.4.7 does not include those versions of the functions

File: dbldatagen/schema_parser.py [hunk 4]
```diff
@@ -6,83 +6,259 @@
...
+            # handle string types of the form "STRING", "char(10)" etc
+            string_keyword = pp.oneOf(["string", "nvarchar", "varchar", "char"], caseless=True, asKeyword=True)
+            string_keyword.setParseAction(lambda s: "string")
+            string_type_expr = string_keyword + pp.Optional(lbracket + number + rbracket)
+
+            # handle decimal types of the form "decimal(10)", "real", "numeric(10,3)"
+            decimal_keyword = pp.MatchFirst(
+                [pp.CaselessKeyword("decimal"), pp.CaselessKeyword("dec"), pp.CaselessKeyword("number"),
+                 pp.CaselessKeyword("numeric")])
+            decimal_keyword.setParseAction(lambda s, loc, tok: "decimal")
```
[the remark this answers is not in the document]
Comment (ronanstokes-db) [reply dbldatagen/schema_parser.py]: should be fixed

**Review by howardwu-db — APPROVED** [review]

**Pull request merged**